vine/std/numeric/Nat.vi

```diff
@@ -0,0 +1,301 @@
+
+use std::ops::arithmetic::{Add, Div, Mul, Rem, Sub};
+use std::ops::{Cast, comparison::{Eq, Ord}};
+use std::data::Iterator::Fused;
+use std::debug::Show;
+
+pub struct* Nat(List[N32]);
+
+pub mod Nat {
+  pub const zero: Nat = Nat([]);
+  pub const one: Nat = Nat([1]);
+
+  pub fn from_parts(parts: List[N32]) -> Nat {
+    parts.trim_back_where(fn* (&n: &N32) { n == 0 });
+    Nat(parts)
+  }
+
+  pub fn .to_parts(Nat(parts)) -> List[N32] {
+    parts
+  }
+
+  pub impl add: Add[Nat, Nat, Nat] {
+    fn add(Nat(a), Nat(b)) -> Nat {
+      let result = [];
+      let carry = 0;
+      let remaining = loop {
+        match (a.pop_front(), b.pop_front()) {
+          (Some(a_num), Some(b_num)) {
+            let lo = a_num + b_num;
+            let sum = lo + carry;
+            carry = N32::add_high(a_num, b_num) + N32::add_high(sum, carry);
+            result.push_back(sum);
+            continue;
+          }
+          (None, Some(b_num)) {
+            result.push_back(b_num + carry);
+            carry = N32::add_high(b_num, carry);
+            b
+          }
+          (Some(a_num), None) {
+            result.push_back(a_num + carry);
+            carry = N32::add_high(a_num, carry);
+            a
+          }
+          (None, None) { [] }
+        }
+      };
+      loop {
+        when {
+          carry == 0 {
+            result ++= remaining;
+            break;
+          }
+          remaining.pop_front() is Some(num) {
+            let sum = num + carry;
+            carry = N32::add_high(num, carry);
+            result.push_back(sum);
+            continue;
+          }
+          _ {
+            result.push_back(carry);
+            break;
+          }
+        }
+      }
+      Nat(result)
+    }
+  }
+
+  pub impl sub: Sub[Nat, Nat, Nat] {
+    fn sub(Nat(a), Nat(b)) -> Nat {
+      let result = [];
+      let borrow = false;
+      let ~trim = true;
+      let remaining = loop {
+        match (a.pop_front(), b.pop_front()) {
+          (Some(a_num), Some(b_num)) {
+            let diff = a_num - b_num - borrow as N32;
+            borrow = if borrow {
+              a_num <= b_num
+            } else {
+              a_num < b_num
+            };
+
+            if !~trim {
+              result.push_back(diff);
+            }
+            if diff != 0 {
+              ~trim = false;
+            }
```

Comment:
This is done a few times; I think it would be nice for this to be extracted to a `let fn`